### ISSUE TITLE
upsert into synchronized table fails

### ISSUE THREAD

**Issue by tylerbrandt** [groue/GRDB.swift#1390]
### What did you do?

I defined a table `content` with columns "id" primary key and "text", then a virtual table `fts` which mirrors those columns using [`synchronize(withTable:)`](https://swiftpackageindex.com/groue/grdb.swift/v6.15.0/documentation/grdb/fts5tabledefinition/synchronize(withtable:)), then tried to execute [`upsert`](https://swiftpackageindex.com/groue/grdb.swift/v6.15.0/documentation/grdb/record/upsert(_:)) on the `Content` type (conforming to `PersistableRecord`).

Snippet from the playground reproducible environment:
```swift
try dbQueue.inDatabase({ db in
  try db.create(table: "content", body: { t in
    t.column("id", .text).primaryKey()
    t.column("text", .text)
  })
  
  try db.create(virtualTable: "fts", using: FTS5()) { t in
    t.tokenizer = .porter()
    t.synchronize(withTable: "content")
    t.column("id").notIndexed()
    t.column("text")
  }
})

struct Content: Codable, Hashable, PersistableRecord {
  var id: String
  var text: String
}

do {
  try dbQueue.write { db in
    try Content(id: "foobar", text: "baz").upsert(db)
  }
} catch {
  assertionFailure("Failed to upsert: \(error)")
}
```

### What did you expect to happen?

The row would be added to both `content` and `fts`

### What happened instead?

The transaction aborted with error 4 after running these statements
```
SQL> BEGIN DEFERRED TRANSACTION
SQL> PRAGMA database_list
SQL> SELECT type, name, tbl_name, sql FROM sqlite_master
SQL> PRAGMA main.table_xinfo("content")
SQL> PRAGMA main.table_list("content")
SQL> INSERT INTO "content" ("id", "text") VALUES (?,?) ON CONFLICT DO UPDATE SET "text" = "excluded"."text" RETURNING "rowid"
SQL> -- PRAGMA 'main'.data_version
SQL> -- TRIGGER __fts_ai
SQL> -- INSERT INTO "fts"("rowid", "id", "text") VALUES (new."rowid", new."id", new."text")
SQL> -- REPLACE INTO 'main'.'fts_docsize' VALUES(?,?)
SQL> REPLACE INTO 'main'.'fts_data'(id, block) VALUES(?,?)
SQL> PRAGMA 'main'.data_version
__lldb_expr_26/Content.playground:41: Fatal error: Failed to upsert: SQLite error 4: Transaction was aborted
```

### Environment

**GRDB flavor(s):** (GRDB, SQLCipher, Custom SQLite build?) GRDB
**GRDB version:** 6.15.0
**Installation method:** (CocoaPods, SPM, manual?) SPM
**Xcode version:** 14.3.1
**Swift version:** Not sure, as I ran it in a playground, but I think 5.7
**Platform(s) running GRDB:** (iOS, macOS, watchOS?) 
**macOS version running Xcode:** 13.5 beta

### Demo Project

https://github.com/tylerbrandt/grdb-fts-upsert

**Comment by groue** [comment]
Hello @tylerbrandt,

Thanks for the report, and the demo project. It looks like this one will need some time to be fully understood. Please stay tuned.

Are you stuck?

**Comment by tylerbrandt** [comment]
Hi @groue, thanks for the quick reply! 

I can work around it; in fact it turns out I can use `save` instead of `upsert` for my particular use case (found by accident), which runs two queries to insert but that should be ok. Seems like the particular issue with the `synchronized` triggers only happens in the case of the `ON CONFLICT DO UPDATE SET` query.

Another thing I discovered is that if the table is _not empty_ the `upsert` works fine; this runs without errors, but if you remove the first `save` call, or change it to `upsert`, it fails:
```
try dbQueue.write { db in
  try Content(id: "one", text: "foo").save(db)
  print("Saved")
  try Content(id: "two", text: "bar").upsert(db)
  print("Upserted")
}
```
I'm not familiar enough with the guts of FTS to know if there are cases other than the empty table that might trigger this, and it may very well be an issue with SQLite itself, since the triggers added by `grdb` seem pretty simple.

**Comment by tylerbrandt** [comment]
Good question. I modified the script to omit the "SQL>" and append a ";" to each (expanded) traced line in order to get the "log" of commands. For whatever reason when I run those same SQL commands in the command line it seems to work fine:
```
> sqlite3
SQLite version 3.39.5 2022-10-14 20:58:05
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
sqlite> SELECT * FROM sqlite_master LIMIT 1;
sqlite> CREATE TABLE "content" ("id" TEXT PRIMARY KEY, "text" TEXT);
sqlite> SELECT QUOTE('porter');
'porter'
sqlite> SELECT QUOTE('''porter''');
'''porter'''
sqlite> SELECT QUOTE('content');
'content'
sqlite> PRAGMA database_list;
0|main|
sqlite> SELECT type, name, tbl_name, sql FROM sqlite_master;
table|content|content|CREATE TABLE "content" ("id" TEXT PRIMARY KEY, "text" TEXT)
index|sqlite_autoindex_content_1|content|
sqlite> PRAGMA main.table_xinfo("content");
0|id|TEXT|0||1|0
1|text|TEXT|0||0|0
sqlite> PRAGMA main.table_list("content");
main|content|table|2|0|0
sqlite> BEGIN DEFERRED TRANSACTION;
sqlite> CREATE VIRTUAL TABLE "fts" USING fts5(id UNINDEXED, text, tokenize='''porter''', content='content');
sqlite> -- CREATE TABLE 'main'.'fts_data'(id INTEGER PRIMARY KEY, block BLOB);
sqlite> -- CREATE TABLE 'main'.'fts_idx'(segid, term, pgno, PRIMARY KEY(segid, term)) WITHOUT ROWID;
sqlite> -- REPLACE INTO 'main'.'fts_data'(id, block) VALUES(?,?);
sqlite> -- REPLACE INTO 'main'.'fts_data'(id, block) VALUES(?,?);
sqlite> -- CREATE TABLE 'main'.'fts_docsize'(id INTEGER PRIMARY KEY, sz BLOB);
sqlite> -- CREATE TABLE 'main'.'fts_config'(k PRIMARY KEY, v) WITHOUT ROWID;
sqlite> -- REPLACE INTO 'main'.'fts_config' VALUES(?,?);
sqlite> -- PRAGMA 'main'.data_version;
sqlite> -- SELECT k, v FROM 'main'.'fts_config';
sqlite> PRAGMA database_list;
0|main|
sqlite> SELECT type, name, tbl_name, sql FROM sqlite_master;
table|content|content|CREATE TABLE "content" ("id" TEXT PRIMARY KEY, "text" TEXT)
index|sqlite_autoindex_content_1|content|
table|fts|fts|CREATE VIRTUAL TABLE "fts" USING fts5(id UNINDEXED, text, tokenize='''porter''', content='content')
table|fts_data|fts_data|CREATE TABLE 'fts_data'(id INTEGER PRIMARY KEY, block BLOB)
table|fts_idx|fts_idx|CREATE TABLE 'fts_idx'(segid, term, pgno, PRIMARY KEY(segid, term)) WITHOUT ROWID
table|fts_docsize|fts_docsize|CREATE TABLE 'fts_docsize'(id INTEGER PRIMARY KEY, sz BLOB)
table|fts_config|fts_config|CREATE TABLE 'fts_config'(k PRIMARY KEY, v) WITHOUT ROWID
sqlite> PRAGMA main.table_xinfo("content");
0|id|TEXT|0||1|0
1|text|TEXT|0||0|0
sqlite> PRAGMA main.table_list("content");
main|content|table|2|0|0
sqlite> CREATE TRIGGER "__fts_ai" AFTER INSERT ON "content" BEGIN
   ...>     INSERT INTO "fts"("rowid", "id", "text") VALUES (new."rowid", new."id", new."text");
   ...> END;
sqlite> CREATE TRIGGER "__fts_ad" AFTER DELETE ON "content" BEGIN
   ...>     INSERT INTO "fts"("fts", "rowid", "id", "text") VALUES('delete', old."rowid", old."id", old."text");
   ...> END;
sqlite> CREATE TRIGGER "__fts_au" AFTER UPDATE ON "content" BEGIN
   ...>     INSERT INTO "fts"("fts", "rowid", "id", "text") VALUES('delete', old."rowid", old."id", old."text");
   ...>     INSERT INTO "fts"("rowid", "id", "text") VALUES (new."rowid", new."id", new."text");
   ...> END;
sqlite> INSERT INTO "fts"("fts") VALUES('rebuild');
sqlite> -- DELETE FROM 'main'.'fts_data';
sqlite> -- DELETE FROM 'main'.'fts_idx';
sqlite> -- DELETE FROM 'main'.'fts_docsize';
sqlite> -- REPLACE INTO 'main'.'fts_data'(id, block) VALUES(?,?);
sqlite> -- REPLACE INTO 'main'.'fts_data'(id, block) VALUES(?,?);
sqlite> -- REPLACE INTO 'main'.'fts_config' VALUES(?,?);
sqlite> -- SELECT T.'rowid', T.'id', T.'text' FROM 'main'.'content' AS T;
sqlite> -- REPLACE INTO 'main'.'fts_data'(id, block) VALUES(?,?);
sqlite> COMMIT TRANSACTION;
sqlite> -- REPLACE INTO 'main'.'fts_data'(id, block) VALUES(?,?);
sqlite> BEGIN DEFERRED TRANSACTION;
sqlite> PRAGMA database_list;
0|main|
sqlite> SELECT type, name, tbl_name, sql FROM sqlite_master;
table|content|content|CREATE TABLE "content" ("id" TEXT PRIMARY KEY, "text" TEXT)
index|sqlite_autoindex_content_1|content|
table|fts|fts|CREATE VIRTUAL TABLE "fts" USING fts5(id UNINDEXED, text, tokenize='''porter''', content='content')
table|fts_data|fts_data|CREATE TABLE 'fts_data'(id INTEGER PRIMARY KEY, block BLOB)
table|fts_idx|fts_idx|CREATE TABLE 'fts_idx'(segid, term, pgno, PRIMARY KEY(segid, term)) WITHOUT ROWID
table|fts_docsize|fts_docsize|CREATE TABLE 'fts_docsize'(id INTEGER PRIMARY KEY, sz BLOB)
table|fts_config|fts_config|CREATE TABLE 'fts_config'(k PRIMARY KEY, v) WITHOUT ROWID
trigger|__fts_ai|content|CREATE TRIGGER "__fts_ai" AFTER INSERT ON "content" BEGIN
    INSERT INTO "fts"("rowid", "id", "text") VALUES (new."rowid", new."id", new."text");
END
trigger|__fts_ad|content|CREATE TRIGGER "__fts_ad" AFTER DELETE ON "content" BEGIN
    INSERT INTO "fts"("fts", "rowid", "id", "text") VALUES('delete', old."rowid", old."id", old."text");
END
trigger|__fts_au|content|CREATE TRIGGER "__fts_au" AFTER UPDATE ON "content" BEGIN
    INSERT INTO "fts"("fts", "rowid", "id", "text") VALUES('delete', old."rowid", old."id", old."text");
    INSERT INTO "fts"("rowid", "id", "text") VALUES (new."rowid", new."id", new."text");
END
sqlite> PRAGMA main.table_xinfo("content");
0|id|TEXT|0||1|0
1|text|TEXT|0||0|0
sqlite> PRAGMA main.table_list("content");
main|content|table|2|0|0
sqlite> INSERT INTO "content" ("id", "text") VALUES ('foobar','baz') ON CONFLICT DO UPDATE SET "text" = "excluded"."text" RETURNING "rowid";
1
sqlite> -- PRAGMA 'main'.data_version;
sqlite> INSERT INTO "content" ("id", "text") VALUES ('foobar','baz') ON CONFLICT DO UPDATE SET "text" = "excluded"."text" RETURNING "rowid";
1
sqlite> INSERT INTO "content" ("id", "text") VALUES ('foobar','baz') ON CONFLICT DO UPDATE SET "text" = "excluded"."text" RETURNING "rowid";
1
sqlite> -- REPLACE INTO 'main'.'fts_docsize' VALUES(?,?);
sqlite> REPLACE INTO 'main'.'fts_data'(id, block) VALUES(1,x'010001');
sqlite> PRAGMA 'main'.data_version;
2
```
And afterwards I can query the result:
```
sqlite> commit;
sqlite> select * from content;
foobar|baz
```
Maybe it's something to do with GRDB's embedded DB? Or the `DatabaseQueue` implementation (I noted the same behavior occurred with `DatabasePool` however)?

**Comment by groue** [comment]
@tylerbrandt, I can reproduce the error with your sample code with multiple SQLite versions shipped on macOS and iOS (3.39.5 and 3.36).

But I can't reproduce with the command line tool (3.39.5).

And I can't reproduce with a GRDB + custom SQLite build (3.39.3). I'm disappointed here, because I was planning to better understand the issue by breakpointing inside SQLite.

Still investigating.

**Comment by groue** [comment]
Some update.

The error we see is thrown by GRDB, not by SQLite. GRDB detects that the transaction that wraps the `upsert` was aborted. If I comment out this error, the test still fails, because the `COMMIT` that ends the transaction is rejected by SQLite anyway:

```sql
-- try dbQueue.write { db in
--   try Content(id: "foobar", text: "baz").upsert(db)
-- }

BEGIN DEFERRED TRANSACTION;

-- Aborts the transaction for some reason.
-- This is detected by GRDB: "Transaction was aborted"
INSERT INTO content ...;

-- Not executed when GRBD throws an error in the
-- above statement. Rejected by SQLite anyway:
-- "cannot commit - no transaction is active"
COMMIT TRANSACTION;
```

I'm still unable to spot what is causing the transaction to be aborted, even after disabling most GRDB "decorations" on top of raw SQLite connections. Still investigating.

**Comment by groue** [comment]
OK, the problem has been found, and solved in #1391.

The cause of the issue was that GRDB used to call `sqlite3_step` only once for upsert statements (one step was enough to grab the values requested by the `RETURNING` clause). Unfortunately, this single step wasn't enough for FTS5 synchronization (or triggers, what do I know?) to apply. The upsert was thus not fully applied, and SQLite would perform the safest action in case of such a misuse: abort everything.

The fix was to run `sqlite3_step` until it returns `SQLITE_DONE`.

This wan't easy to find the root cause 😅 

Anyway, thank you very much for the report, @tylerbrandt. Next GRDB version will ship with the fix, and you'll be able to remove your workaround.

**Comment by groue** [comment]
The fix has shipped in v6.15.1

**Comment by tylerbrandt** [comment]
Thank you! I appreciated the explanation as well; I learned a little bit about GRDB and SQLite. ☺️
